// @pinetwork-js/sdk/build/MessageTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageType = void 0;
var MessageType;
(function (MessageType) {
    MessageType["AUTH_ERROR"] = "@pi:app:error:auth";
    MessageType["UNKNOWN_ERROR"] = "@pi:app:error:unknown";
    MessageType["OPEN_APP_CONVERSATION_WITH_ID"] = "@pi:app:conversation:open_with_id";
    MessageType["OPEN_SHARE_DIALOG_ACTION"] = "@pi:app:share_dialog:open";
    MessageType["PREPARE_PAYMENT_FLOW"] = "@pi:app:payments:prepare_payment_flow";
    MessageType["SHOW_PRE_PAYMENT_ERROR"] = "@pi:app:payments:show_pre_payment_error";
    MessageType["START_PAYMENT_FLOW"] = "@pi:app:payments:start_payment_flow";
    MessageType["WAIT_FOR_TRANSACTION"] = "@pi:app:payments:wait_for_transaction";
    MessageType["COMMUNICATION_INFORMATION_REQUEST"] = "@pi:app:sdk:communication_information_request";
    MessageType["OPEN_CONSENT_MODAL"] = "@pi:app:sdk:open_consent_modal";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
